server/src/main/java/org/apache/druid/metadata/IndexerSQLMetadataStorageCoordinator.java

```diff
@@ -1423,4 +1424,40 @@ public boolean insertDataSourceMetadata(String dataSource, DataSourceMetadata me
             .execute()
     );
   }
+
+  @Override
+  public int removeDataSourceMetadataOlderThan(long timestamp, @Nullable Set<String> excludeDatasources)
+  {
+    DateTime dateTime = DateTimes.utc(timestamp);
+    List<String> datasources = connector.getDBI().withHandle(
+        handle -> handle
+            .createQuery(
+                StringUtils.format(
+                    "SELECT dataSource FROM %1$s WHERE created_date < '%2$s'",
+                    dbTables.getDataSourceTable(),
+                    dateTime.toString()
+                )
+            )
+            .mapTo(String.class)
+            .list()
+    );
+    return connector.getDBI().withHandle(
+        handle -> {
+          final PreparedBatch batch = handle.prepareBatch(
+              StringUtils.format(
+                  "DELETE FROM %1$s WHERE dataSource = :dataSource AND created_date < '%2$s'",
```

Comment:
Why did you choose to build the delete statements one at a time instead of doing a batch delete?

I think we could encapsulate the `excludeDatasources` logic in a where clause of this delete statement instead.

Something like `DELETE FROM datasources where created_date < "date" and datasource not in ("excludeDataSources")`

Reply:
This is to prevent the IN clause being too large

Reply:
Also moved the filtering of the datasources to outside the handle block.